c_common/front_end_common_lib/include/spinn_extra.h

```diff
@@ -0,0 +1,1106 @@
+
+typedef struct {
+    uint load_value;
+    const uint current_value;
+    timer_control_t control;
+    uint interrupt_clear;
+    const timer_interrupt_status_t raw_interrupt_status;
+    const timer_interrupt_status_t masked_interrupt_status;
+    uint background_load_value;
+    uint _dummy;
+} timer_controller_t;
+
+ASSERT_WORD_SIZED(timer_control_t);
+ASSERT_WORD_SIZED(timer_interrupt_status_t);
+
+// ---------------------------------------------------------------------
+// 7. DMA Controller
+
+typedef struct {
+    uint _zeroes : 2;
+    uint length_words : 15;
+    uint : 2;
+    uint direction : 1; // 0 = write to TCM, 1 = write to SDRAM
+    uint crc : 1;
+    uint burst : 3;
+    uint width : 1; // 0 = word, 1 = double-word
+    uint privilege : 1;
+    uint transfer_id : 6;
+} dma_description_t;
+
+typedef struct {
+    uint uncommit : 1;
+    uint abort : 1;
+    uint restart : 1;
+    uint clear_done_int : 1;
+    uint clear_timeout_int : 1;
+    uint clear_write_buffer_int : 1;
+    uint : 26;
+} dma_control_t;
+
+typedef struct {
+    uint transferring : 1;
+    uint paused : 1;
+    uint queued : 1;
+    uint write_buffer_full : 1;
+    uint write_buffer_active : 1;
+    uint : 5;
+    uint transfer_done : 1;
+    uint transfer2_done : 1;
+    uint timeout : 1;
+    uint crc_error : 1;
+    uint tcm_error : 1;
+    uint axi_error : 1;
+    uint user_abort : 1;
+    uint soft_reset : 1;
+    uint : 2;
+    uint write_buffer_error : 1;
+    uint : 3;
+    uint processor_id : 8;
+} dma_status_t;
+
+typedef struct {
+    uint bridge_buffer_enable : 1;
+    uint : 9;
+    uint transfer_done_interrupt : 1;
+    uint transfer2_done_interrupt : 1;
+    uint timeout_interrupt : 1;
+    uint crc_error_interrupt : 1;
+    uint tcm_error_interrupt : 1;
+    uint axi_error_interrupt : 1;
+    uint user_abort_interrupt : 1;
+    uint soft_reset_interrupt : 1;
+    uint unknown1 : 1;
+    uint unknown2 : 1;
+    uint write_buffer_error_interrupt : 1;
+    uint : 10;
+    uint timer : 1;
+} dma_global_control_t;
+
+typedef struct {
+    uint _zeroes : 5;
+    uint value : 5;
+    uint : 22;
+} dma_timeout_t;
+
+typedef struct {
+    uint enable : 1;
+    uint clear : 1;
+    uint : 30;
+} dma_stats_control_t;
+
+typedef struct {
+    const uint _unused1[1];
+    void *sdram_address; // r1
+    void *tcm_address;
+    volatile dma_description_t description;
+    volatile dma_control_t control;
+    const dma_status_t status;
+    volatile dma_global_control_t global_control;
+    const uint crcc;
+    const uint crcr;
+    dma_timeout_t timeout;
+    dma_stats_control_t statistics_control; // r10
+    const uint _unused2[5];
+    const uint statistics[8]; // r16-r23
+    const uint _unused3[41];
+    const void *current_sdram_address; // r65
+    const void *current_tcm_address;
+    const dma_description_t current_description; // r67
+    const uint _unused4[29];
+    uint crc_polynomial[32];
+} dma_t;
+
+ASSERT_WORD_SIZED(dma_description_t);
```

Comment:
checks structs in this section are word aligned.

Reply:
lost interest. but i believe all these structs and values should have a doc associated with them. likely linking to the datasheet pages

Reply:
documented… and there's a link to the [SpiNNaker datasheet](https://spinnakermanchester.github.io/docs/SpiNN2DataShtV202.pdf) (which is the definitive document here) but it's basically just once or twice for the whole file because I'm a bit lazy and don't want to do a massive cut-n-paste in everywhere. (I also don't think the datasheet has internal anchors, so I can only link to the whole document.)